feat(build): Add polyfill-extraction rollup plugin

lobsterkatie opened the pull request:
Both Rollup and Sucrase add polyfills (similar to those in `tslib`) during the build process, to down-compile newer language features and help ESM and CJS modules play nicely together. Unlike with `tslib`, however, there's no option equivalent to tsc's [`importHelpers` option](https://www.typescriptlang.org/tsconfig#importHelpers), which allows those polyfills to be imports from one central place rather than copies of the full code for each necessary polyfill in every file that uses them. Having all that repeated code is a surefire way to wreck our bundle-size progress, so this PR replicates the behavior of `importHelpers` in the context of our new sucrase/rollup builds.

Key changes:

- The polyfills themselves have been added to the `rollup` folder. Since some treeshaking algorithms can only work file by file, each one is in its own file. Further, where possible, they've been rewritten to be a bit more compact. (This rewriting mostly applies to the import helpers.)
- There is now a rollup config just for the polyfills, which both moves them into `@sentry/utils` and compiles them into ESM and CJS varieties.
- In order to have access to the polyfills, all packages which didn't already have `@sentry/utils` as a dependency now do.
- There is now a rollup plugin, which runs after both Sucrase and Rollup do their polyfill injection, which parses the built files and replaces every instance of a polyfill function definition with an equivalent `import` or `require` statement.

A note on file location:

The polyfills themselves live in the rollup folder, and once compiled they end up at the root level of the utils package, in a `jsPolyfills` directory. When other other packages use them, they either import from `'@sentry/utils/jsPolyfills/esm'` or require `'@sentry/utils/jsPolyfills/cjs'`. Both of those choices break pattern for us: We have no other code which lives outside of `packages/*/src`, and we nowhere else do we  import from anywhere other than `@sentry/<package>`. These seemed like acceptable and worthwhile deviations, though, for the following reasons:

- Originals in repo-level `rollup` directory:
    - We didn't write the original polyfills, and they aren't written in TS, so putting them into one of our `src` directories would have both felt kinda wrong and necessitated changes to our build config to handle non-TS files.
    - In the long term, they might well live in a `@sentry/dev-utils`/`@sentry/internal-utils`/`@sentry/whatever` package, but to get things running, for now the `rollup` directory seemed like a good place, since they are inextricably tied up in our rollup build.
- Built files in root level of `@sentry/utils`:
    - Other polyfills live in `@sentry/utils`, and almost all of our packages already depend on it, so it seemed like a good place to import them from.
    - That said, the polyfills are strictly for internal use, and we don't want them to be part of the public API, which they would be if we were to export them directly from `@sentry/utils`. Exporting them from a subdirectory effectively hides them in plain sight - we can get to them easily, but they don't show up anywhere as being part of `@sentry/utils`.

Doing it this way means we will also need to add a step to our prepack script, to make sure they end up in our npm tarball, but that will be handled in a separate PR.

AbhiPrasad commented:
Does this mean we can remove `tslib` as a dependency?

lobsterkatie commented:
> Does this mean we can remove `tslib` as a dependency?

Once we switch over, yes, that's the idea. We're basically making our own `tslib`.

AbhiPrasad commented on rollup/plugins/extractPolyfillsPlugin.js:
```diff
@@ -0,0 +1,208 @@
+  const importSource = literal(
+    isUtilsPackage
+      ? path.relative(path.dirname(currentSourceFile), `../jsPolyfills/${moduleFormat}`)
+      : `@sentry/utils/jsPolyfills/${moduleFormat}`,
```
should we make another package for the polyfills?

lobsterkatie replied:
> In the long term, they might well live in a `@sentry/dev-utils`/`@sentry/internal-utils`/`@sentry/whatever package`, but to get things running, for now the rollup directory seemed like a good place

In the long run I agree that'd be a better place, but I'd say it's okay to leave this as an iterative improvement. What do you think?

Lms24 replied:
I'm +1 on an extra package but it's probably fine if we do it in a follow-up task (whenever).

lobsterkatie replied:
I don't think they warrant a package all on their own, but I did experiment with moving them to a `dev-utils` package. Once I got them there, though, I didn't love that it was then going to become a runtime dependency of `@sentry/utils`, when it really should only be a dev dependency. And as part of that experimentation, I did end up translating them to TS, which was the final remaining blocker to them living directly in `@sentry/utils`, so that is where they ended up.

(The other two original blockers for me were:
- Feeling weird about putting code that wasn't ours in with code that was. This was solved for me by 1) the fact that we now go overboard in terms of attribution, license-copying, etc, and 2) the fact that in fact they aren't exact copies of the originals anymore, if for no other reason than the translation to TS.
- Not wanting them to be part of the public API of `@sentry/utils`. Originally that meant that I was copying them over to a folder outside of `build`, but I realized that their built versions actually can live inside `build` and still be "hidden" (in that they won't show up in intellisense completions, for example) as long as they're not exported from `index.ts`. That then let me put the folder containing the pre-build versions straight into `src/`, which made everything about building and testing them a lot easier.)

AbhiPrasad commented on rollup/jsPolyfills/_asyncNullishCoalesce.js:
```diff
@@ -0,0 +1,4 @@
+export async function _asyncNullishCoalesce(lhs, rhsFn) {
+  // by checking for loose equality to `null`, we catch both `null` and `undefined`
+  return lhs != null ? lhs : rhsFn();
```
It would be good to test these! We can do it later on though.

lobsterkatie replied:
Yes, though to the extent we test either sucrase or rollup at all, we have to trust these functions, as all but the interop ones are verbatim copied from what those packages inject. Still, if we had tests, you're right that we could try to refactor more of them to be more efficient.

(Not sure how much that's possible, but without tests I'm not even going to try to, for instance, refactor the optional chaining function. The interop ones I felt pretty confident about because they're super straightforward (as were most of my refactors, things like assigning directly rather than using `Object.defineProperty`), but for more complicated ones, I'd definitely want tests before I started mucking about.)

I'll throw it onto the backlog.

Lms24 reviewed (commented):
So I took a look at the PR and overall this looks good to me with the asterisk that I'm not entirely comfortable with having our "own" polyfills because to me this seems kind of strange. But then again, I clearly don't have as much context here as you so please take it with a grain of salt. 

I do have a few questions though: 
* Just to double check - there isn't a way around having our own polyfills, right? I was thinking of something like importing them from sucrase/rollup/tslib instead of adding them to our repo where we have to maintain them. In my mind it would be best to just import them and add the `importHelpers` functionality around them. But again, I probably lack a lot of necessary context here.
* The polyfills we're adding to our repo in this PR - are they exact copies of the Sucrase/Rollup polyfilles? 
  * If yes, do we need to acknowledge this in the code files/give credits, etc? 
  * if no, I really think we should have tests for them (would be best either way, I guess)

Anyway, great work and thanks for figuring all of this out!

lobsterkatie commented:
> * Just to double check - there isn't a way around having our own polyfills, right? I was thinking of something like importing them from sucrase/rollup/tslib instead of adding them to our repo where we have to maintain them.

Believe me, I wish I saw a way around it, because I agree, I'd rather leave maintenance/testing/etc to them instead of having to do it ourselves. But it's unfortunately not as simple as importing the helpers, because they don't exist in a straightforwardly importable form. Sucrase [uses a `HelperManager` class](https://github.com/alangpierce/sucrase/blob/265887868966917f3b924ce38dfad01fbab1329f/src/HelperManager.ts#L126), which takes string versions of the functions, changes their names, and then injects them into each file, and Rollup [generates the code itself on the fly](https://github.com/rollup/rollup/blob/f44a3a38b357c2ba80a99aca36cfd30dcfb84476/src/utils/interopHelpers.ts#L142-L165) before injecting it. So the alternative to vendoring the polyfills would be to try to tap into their polyfill-generation machinery, which felt a lot more brittle, as it's much more likely they'd change that than the actual code of the helpers (which are simple, dependent on nothing, and solving a known and fixed problem, all of which means they're likely to be highly stable over time). It also seemed like it'd also be much easier for us to mess that up, compared to just grabbing the helpers in their final form and throwing them into a folder.

> * The polyfills we're adding to our repo in this PR - are they exact copies of the Sucrase/Rollup polyfilles?  

In the case of everything other than the iterop helpers, yes, they're the same as what's emitted by sucrase (the only ones emitted by rollup are the interop ones). As far as the interop ones go, [here](https://rollupjs.org/guide/en/#outputinterop) are the originals (we use either the `esmodule` option or the `auto` option, depending on the package's current `esModuleInterop` value). The versions included in this PR are definitely simplified, but (I claim) not in a way which changes the outcome. (For example, I turned `var n = Object.create(null)` into `var n = {}`, and I mimicked the behavior of [setting `output.externalLiveBindings` to `false`](https://rollupjs.org/guide/en/#outputexternallivebindings) and [`output.freeze` to `false`](https://rollupjs.org/guide/en/#outputfreeze).)

>   * If yes, do we need to acknowledge this in the code files/give credits, etc?

From the first line of the README distributed with the polyfills: "This is a collection of syntax polyfills either copied directly from or heavily inspired by those used by [Rollup](https://github.com/rollup/rollup) and [Sucrase](https://github.com/alangpierce/sucrase)."

>   * if no, I really think we should have tests for them (would be best either way, I guess)

For those interop ones, I'm happy to work on tests now if it would up our overall confidence level in making this change. I'll update here once I've added some.
